### PR TITLE
refactor: synthetic node

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
#71 minus the fuzz test changes.

Performance and conformance tests now use the SyntheticNode. Includes some API improvements.

### CODE REVIEW

**Review by niklaslong — APPROVED** [review]
Left a few comments/thoughts for your consideration. Great work on this 🎉

File: src/helpers/synthetic_peers.rs [hunk 1]
```diff
@@ -105,13 +183,36 @@ impl SyntheticNode {
         self.inner_node.node().known_peers()
     }
 
+    /// Waits until the node has at least one connection, and
+    /// returns its SocketAddr
+    pub async fn wait_for_connection(&self) -> SocketAddr {
+        const SLEEP: Duration = Duration::from_millis(10);
+        loop {
+            {
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/helpers/synthetic_peers.rs:191]: Ah nice. I think I only arrived at the `if let` at the end - I remember trying various things which always resulted in the lock not living long enough to access its insides. But I obviously didn't try it with de-structuring..

File: src/helpers/synthetic_peers.rs [hunk 2]
```diff
@@ -156,25 +261,102 @@ impl SyntheticNode {
...
 
+    /// Sends Ping(nonce), and expects Pong(nonce) as a reply.
+    ///
+    /// Uses polling to check that connection is still alive.
+    ///
+    /// Errors if:
+    ///     - a non-Pong(nonce) message is received
+    ///     - timeout expires
+    ///     - connection breaks
+    pub async fn ping_pong_timeout(
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/helpers/synthetic_peers.rs:272]: I'll add that to my refactor todo.

File: src/helpers/synthetic_peers.rs [hunk 3]
```diff
@@ -315,6 +500,26 @@ impl Handshaking for InnerNode {
...
+
+                let version = Message::read_from_stream(conn.reader()).await?;
+                assert_matches!(version, Message::Version(..));
+            }
+            (Some(Handshake::VersionOnly), ConnectionSide::Responder) => {
+                // Receive and send Version.
+                let version = Message::read_from_stream(conn.reader()).await?;
+                assert_matches!(version, Message::Version(..));
+
+                // FIXME: send the node addr with its listener port, not the ephemeral port
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/helpers/synthetic_peers.rs:516]: We should have access to `self.listener_addr()` at this point?

[the remark this answers is not in the document]
Comment (niklaslong) [reply src/helpers/synthetic_peers.rs:516]: Indeed 👍

File: src/tests/performance/blocks.rs [hunk 4]
```diff
@@ -147,27 +120,34 @@ async fn getdata_blocks_latency() {
                 .collect::<VecDeque<_>>();
 
             peer_handles.push(tokio::spawn(async move {
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/tests/performance/blocks.rs:122]: I was thinking we can have a `Trait` which defines how to react to a message of type `X`. We can provide our current implementations, and then provide `fn` to allow the user to set these. A `Trait` also lets the user store additional context / make their `struct` more complicated if they wish.

The only current example I know of in our code base is where we disable `ping` responding, and then re-enable it a few lines later.

Overall I think its a nice solution; but we don't actually don't require anything like it yet

**Comment by Mirko-von-Leipzig** [comment]
I've updated with some of the review changes.

If you could [double-check this part](https://github.com/eqlabs/ziggurat/pull/72#discussion_r656252965). According to the protocol the order should be `other-node`, `our-node` so I think this is correct? (It was the other way around before)

**Comment by niklaslong** [comment]
LGTM 👍 
> If you could double-check this part
This just links to this PR but I'm assuming you mean the order of the addrs in the version message? If so it was incorrect in the handshake code, the order is indeed `other-node`, `our node`.

**Pull request merged**